File: scripts/data-separation-audit.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const DEFAULT_REPORT_FILE = path.join(
  ROOT,
  "data",
  "output",
  "databases",
  "polyglot-default",
  "analysis",
  "data_separation_audit_report.json"
);

const DEFAULT_SCAN_DIRS = Object.freeze(["main", "app", "renderer", "brain", "scripts"]);
const DEFAULT_IGNORE_DIRS = Object.freeze(["node_modules", ".git", "dist", "data/output", "to-do", "artifacts"]);
const TOKEN_HINTS = Object.freeze([
  "GROUP",
  "LABEL",
  "ALIAS",
  "MAP",
  "REGISTRY",
  "SPEC",
  "KEY",
  "DEFAULT",
  "CONFIG",
  "PATTERN"
]);

function nowIso() {
  return new Date().toISOString();
}

function ensureDir(dirPath) {
  fs.mkdirSync(dirPath, { recursive: true });
}

function normalizePath(value) {
  return String(value || "").replace(/\\/g, "/");
}

function toRelativePath(filePath) {
  return normalizePath(path.relative(ROOT, filePath));
}

function isIgnoredPath(relativePath) {
  const rel = normalizePath(relativePath);
  return DEFAULT_IGNORE_DIRS.some((ignore) => rel === ignore || rel.startsWith(`${ignore}/`));
}

function parseArgs(argv) {
  const args = {
    reportFile: DEFAULT_REPORT_FILE,
    enforce: false
  };

  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--report-file") {
      args.reportFile = path.resolve(process.cwd(), String(argv[index + 1] || ""));
      index += 1;
      continue;
    }
    if (token === "--enforce") {
      args.enforce = true;
      continue;
    }
    if (token === "--help" || token === "-h") {
      printHelpAndExit(0);
    }
    throw new Error(`unknown argument: ${token}`);
  }

  return args;
}

function printHelpAndExit(code) {
  const help = [
    "data-separation-audit",
    "",
    "Usage:",
    "  npm run audit:data-separation -- [options]",
    "",
    "Options:",
    "  --report-file <path>   Output report path",
    "  --enforce              Exit non-zero when required-separation candidates exist",
    "  --help                 Show help"
  ].join("\n");
  process.stdout.write(`${help}\n`);
  process.exit(code);
}

function collectJsFiles() {
  const files = [];
  const stack = DEFAULT_SCAN_DIRS
    .map((entry) => path.join(ROOT, entry))
    .filter((entry) => fs.existsSync(entry) && fs.statSync(entry).isDirectory());

  while (stack.length > 0) {
    const current = stack.pop();
    const relCurrent = toRelativePath(current);
    if (isIgnoredPath(relCurrent)) {
      continue;
    }
    const entries = fs.readdirSync(current, { withFileTypes: true });
    entries.forEach((entry) => {
      const absolute = path.join(current, entry.name);
      const relative = toRelativePath(absolute);
      if (isIgnoredPath(relative)) {
        return;
      }
      if (entry.isDirectory()) {
        stack.push(absolute);
        return;
      }
      if (!entry.isFile()) {
        return;
      }
      if (!relative.endsWith(".js")) {
        return;
      }
      files.push(absolute);
    });
  }

  files.sort((left, right) => toRelativePath(left).localeCompare(toRelativePath(right)));
  return files;
}

function classifyConstant(name) {
  const upper = String(name || "").toUpperCase();
  const requiredToken = TOKEN_HINTS.find((token) => upper.includes(token)) || "";
  if (requiredToken) {
    return {
      separationRequired: true,
      reason: `name contains ${requiredToken}`
    };
  }
  return {
    separationRequired: false,
    reason: "non-governed constant name"
  };
}

function detectCandidatesInFile(filePath) {
  const content = fs.readFileSync(filePath, "utf8");
  const lines = content.split(/\r?\n/);
  const candidates = [];
  const regex = /^\s*const\s+([A-Z0-9_]+)\s*=\s*(?:Object\.freeze\()?(?:\{|\[)/;

  lines.forEach((line, index) => {
    const match = line.match(regex);
    if (!match) {
      return;
    }
    const name = String(match[1] || "");
    const classification = classifyConstant(name);
    candidates.push({
      file: toRelativePath(filePath),
      line: index + 1,
      constant: name,
      separation_required: classification.separationRequired,
      reason: classification.reason
    });
  });

  return candidates;
}

function summarizeByPath(candidates) {
  const grouped = {};
  candidates.forEach((candidate) => {
    const file = candidate.file;
    if (!grouped[file]) {
      grouped[file] = {
        file,
        candidate_count: 0,
        required_count: 0
      };
    }
    grouped[file].candidate_count += 1;
    if (candidate.separation_required) {
      grouped[file].required_count += 1;
    }
  });
  return Object.values(grouped).sort((left, right) => {
    if (right.required_count !== left.required_count) {
      return right.required_count - left.required_count;
    }
    return left.file.localeCompare(right.file);
  });
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const files = collectJsFiles();
  const candidates = files.flatMap((filePath) => detectCandidatesInFile(filePath));
  const required = candidates.filter((candidate) => candidate.separation_required);
  const summaryByFile = summarizeByPath(candidates);

  const report = {
    generated_at: nowIso(),
    project_root: ROOT,
    scan_dirs: DEFAULT_SCAN_DIRS,
    ignored_dirs: DEFAULT_IGNORE_DIRS,
    counts: {
      files_scanned: files.length,
      candidate_total: candidates.length,
      separation_required_total: required.length
    },
    highest_priority_files: summaryByFile.slice(0, 25),
    candidates
  };

  ensureDir(path.dirname(args.reportFile));
  fs.writeFileSync(args.reportFile, `${JSON.stringify(report, null, 2)}\n`, "utf8");

  const result = {
    report_file: args.reportFile,
    files_scanned: files.length,
    candidate_total: candidates.length,
    separation_required_total: required.length,
    enforce: args.enforce
  };

  process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);

  if (args.enforce && required.length > 0) {
    process.exit(2);
  }
}

try {
  main();
} catch (error) {
  process.stderr.write(`data-separation-audit failed: ${error.message}\n`);
  process.exit(1);
}

File: scripts/polyglot/equivalents/python/_native/data_separation_audit.py
#!/usr/bin/env python3
"""Native Python implementation for scripts/data-separation-audit.js."""

from __future__ import annotations

import json
import pathlib
import re
import sys
from datetime import UTC, datetime
from typing import Any

from _common import find_repo_root, normalize_path, write_text_file_robust

ROOT = find_repo_root(pathlib.Path(__file__))
DEFAULT_REPORT_FILE = ROOT / "data" / "output" / "databases" / "polyglot-default" / "analysis" / "data_separation_audit_report.json"
DEFAULT_SCAN_DIRS = ("main", "app", "renderer", "brain", "scripts")
DEFAULT_IGNORE_DIRS = ("node_modules", ".git", "dist", "data/output", "to-do", "artifacts")
TOKEN_HINTS = ("GROUP", "LABEL", "ALIAS", "MAP", "REGISTRY", "SPEC", "KEY", "DEFAULT", "CONFIG", "PATTERN")
CONST_PATTERN = re.compile(r"^\s*const\s+([A-Z0-9_]+)\s*=\s*(?:Object\.freeze\()?(?:\{|\[)")


def now_iso() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def to_relative_path(file_path: pathlib.Path) -> str:
    return normalize_path(file_path.relative_to(ROOT))


def is_ignored_path(relative_path: str) -> bool:
    rel = normalize_path(relative_path)
    return any(rel == ignore_dir or rel.startswith(f"{ignore_dir}/") for ignore_dir in DEFAULT_IGNORE_DIRS)


def parse_args(argv: list[str]) -> dict[str, Any]:
    args: dict[str, Any] = {
        "report_file": DEFAULT_REPORT_FILE,
        "enforce": False,
    }
    index = 0
    while index < len(argv):
        token = str(argv[index] or "")
        if token == "--report-file":
            if index + 1 >= len(argv):
                raise RuntimeError("--report-file requires a path")
            args["report_file"] = (pathlib.Path.cwd() / str(argv[index + 1] or "")).resolve()
            index += 2
            continue
        if token == "--enforce":
            args["enforce"] = True
            index += 1
            continue
        if token in {"--help", "-h"}:
            sys.stdout.write(
                "\n".join(
                    [
                        "data-separation-audit",
                        "",
                        "Usage:",
                        "  npm run audit:data-separation -- [options]",
                        "",
                        "Options:",
                        "  --report-file <path>   Output report path",
                        "  --enforce              Exit non-zero when required-separation candidates exist",
                        "  --help                 Show help",
                    ]
                )
                + "\n"
            )
            raise SystemExit(0)
        raise RuntimeError(f"unknown argument: {token}")
    return args


def collect_js_files() -> list[pathlib.Path]:
    files: list[pathlib.Path] = []
    stack = [ROOT / scan_dir for scan_dir in DEFAULT_SCAN_DIRS if (ROOT / scan_dir).is_dir()]
    while stack:
        current = stack.pop()
        rel_current = to_relative_path(current)
        if is_ignored_path(rel_current):
            continue
        for entry in sorted(current.iterdir(), key=lambda item: str(item)):
            relative = to_relative_path(entry)
            if is_ignored_path(relative):
                continue
            if entry.is_dir():
                stack.append(entry)
                continue
            if entry.is_file() and entry.suffix == ".js":
                files.append(entry)
    return sorted(files, key=lambda item: to_relative_path(item))


def classify_constant(name: str) -> dict[str, Any]:
    upper = str(name or "").upper()
    required_token = next((token for token in TOKEN_HINTS if token in upper), "")
    if required_token:
        return {
            "separation_required": True,
            "reason": f"name contains {required_token}",
        }
    return {
        "separation_required": False,
        "reason": "non-governed constant name",
    }


def detect_candidates_in_file(file_path: pathlib.Path) -> list[dict[str, Any]]:
    candidates: list[dict[str, Any]] = []
    for index, line in enumerate(file_path.read_text(encoding="utf8").splitlines(), start=1):
        match = CONST_PATTERN.match(line)
        if not match:
            continue
        constant_name = str(match.group(1) or "")
        classification = classify_constant(constant_name)
        candidates.append(
            {
                "file": to_relative_path(file_path),
                "line": index,
                "constant": constant_name,
                "separation_required": classification["separation_required"],
                "reason": classification["reason"],
            }
        )
    return candidates


def summarize_by_path(candidates: list[dict[str, Any]]) -> list[dict[str, Any]]:
    grouped: dict[str, dict[str, Any]] = {}
    for candidate in candidates:
        file_name = str(candidate.get("file") or "")
        if file_name not in grouped:
            grouped[file_name] = {
                "file": file_name,
                "candidate_count": 0,
                "required_count": 0,
            }
        grouped[file_name]["candidate_count"] += 1
        if candidate.get("separation_required") is True:
            grouped[file_name]["required_count"] += 1
    return sorted(
        grouped.values(),
        key=lambda entry: (-int(entry.get("required_count") or 0), str(entry.get("file") or "")),
    )


def main(argv: list[str] | None = None) -> int:
    args = parse_args(list(argv or []))
    files = collect_js_files()
    candidates = [candidate for file_path in files for candidate in detect_candidates_in_file(file_path)]
    required = [candidate for candidate in candidates if candidate.get("separation_required") is True]
    summary_by_file = summarize_by_path(candidates)

    report = {
        "generated_at": now_iso(),
        "project_root": str(ROOT),
        "scan_dirs": list(DEFAULT_SCAN_DIRS),
        "ignored_dirs": list(DEFAULT_IGNORE_DIRS),
        "counts": {
            "files_scanned": len(files),
            "candidate_total": len(candidates),
            "separation_required_total": len(required),
        },
        "highest_priority_files": summary_by_file[:25],
        "candidates": candidates,
    }
    write_text_file_robust(pathlib.Path(args["report_file"]), f"{json.dumps(report, indent=2)}\n")

    result = {
        "report_file": str(args["report_file"]),
        "files_scanned": len(files),
        "candidate_total": len(candidates),
        "separation_required_total": len(required),
        "enforce": bool(args["enforce"]),
    }
    sys.stdout.write(f"{json.dumps(result, indent=2)}\n")
    if args["enforce"] and required:
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
